File: app/controllers/translate_controller.rb
require 'httparty'

class TranslateController < ApplicationController
  def translate
    text = params[:text]
    target_lang = 'EN'
    auth_key = ENV['AUTH_KEY']

    url = 'https://api-free.deepl.com/v2/translate'
    headers = {
      'Authorization': "DeepL-Auth-Key #{auth_key}",
      'Content-Type': 'application/x-www-form-urlencoded'
    }
    body = {
      'text': text,
      'target_lang': target_lang
    }

    response = HTTParty.post(url, headers: headers, body: body)
    if response.success?
      render json: { translation: response['translations'][0]['text'] }, status: :ok
    else
      render json: { error: 'Error fetching translation' }, status: :unprocessable_entity
    end
  end

  def reload_i18n
    I18n.reload!
    redirect_to root_path, notice: 'I18n reloaded'
  end
end

File: app/javascript/controllers/translate_controller.js
import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="translate"
export default class extends Controller {
  static targets = ['input', 'output']

  async translateInput(e) {
    e.preventDefault()

    const text = this.inputTarget.value;
    const targetLang = 'EN';

    let translated = await this.translateText(text, targetLang);
    this.outputTarget.value = translated
  }

  async translateText(text, targetLang) {
    const url = '/translate';
    const requestOptions = this.buildRequestOptions(text, targetLang);

    try {
      const response = await fetch(url, requestOptions);
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      const result = await response.json();
      // console.log(JSON.stringify(result));
      const translatedText = result.translation;
      return translatedText;
    } catch (error) {
      console.error('Error fetching translation:', error);
    }
  }

  buildRequestOptions(text, targetLang) {
    const headers = new Headers({
      'Content-Type': 'application/x-www-form-urlencoded'
    });

    const csrfToken = this.getCSRFToken();
    const data = new URLSearchParams();
    data.append('text', text);
    data.append('target_lang', targetLang);
    data.append('authenticity_token', csrfToken);

    return {
      method: 'POST',
      headers: headers,
      body: data
    };
  }

  getCSRFToken() {
    return document.querySelector('meta[name="csrf-token"]').content;
  }
}

